### PR TITLE
Use ${vendor}${year}-${pad}${core}c-${pad}${ram}r product notation

### PR DESCRIPTION
**Pull request by fdr**
Use ${vendor}${year}-${pad}${core}c-${pad}${ram}r product notation

Previously, we used AWS-style names to convey a general idea to those
receiving a demonstration, but there are many problems with AWS-style
names that we don't need to repeat. In particular, they convey nothing
useful about the instance and sort badly, sending me off to the manual
with regularity for simple matters.

The results are like this:

    amd19-0.3c-1.3r
    amd19-0.3c-2.6r
    amd19-1c-4r
    amd19-1c-8r
    amd19-2c-8r
    amd19-2c-t16r
    amd19-4c-t16r
    amd19-4c-t32r
    amd19-8c-t32r
    amd19-8c-t64r
    amd19-t16c-t64r
    amd19-t16c-u128r
    amd19-t32c-u128r
    amd19-t32c-u256r
    amd19-t64c-u256r
    amd19-t64c-u512r
    amd19-u128c-u512r
    amd19-u128c-v1024r

A string like `amd19-2c-8r` means: an AMD chip, released in 2019, two
cores, eight gigabytes of ram.

A string like `amd19-u128c-u512r` means: an AMD chip, released in
2019, 128 cores, 512 gigabytes of ram.  Note the `u` character has no
meaning: it's only for sort assistance, explained in more detail
below.

A string like `amd19-0.3c-1.3r` is intended to be for a third of a
core, a product we don't support yet, but I wanted to include how it
might look.

The meaning of each part:

`c` is cores, `r` is gigabytes of ram.  We could omit these to be
positional like `amd19-2-4`, but I thought it'd invite errors as
people would swap the meaning of the two positions.

the `t` and `u`, `v` bytes seen with two, three, and four digit
numbers respectively are used to help the product strings sort.  They
have no other meaning.  No such sort assist byte is necessary for
decimalized numbers or single-digit numbers.

This is the strangest part of the design, introducing a meaningless
character to assist sorts.  I'm not convinced of it, but bad sorts on
product names have been a pebble in my shoe for a long time.  Normally
this is accomplished by zero-padding, but this would result in long
and hard to type product strings for small products, which tend to be
the most commonly used, e.g. a hypothetical `amd19-002c-0008r`.  I
thought this was intolerable.

== Comparisons to other vendors

In the form of shell expansion with a default: `${symbol:example}`

GCP is of the form:

    ${obscure-microarchitecture-string:t2d}-${ram-ratio-string:standard}-${cores}

Example: `t2d.standard-4`

Azure is of the form:

    Standard_${ram-ratio+manufacturer-char:D}#{vcpu:64}${other-char-modifiers:s}_${microarch_version:v5}

Example: `Standard_D64ds_v4`

I don't know why the `Standard_` shows up so often without variation
in Azure.  Another annoying quirk is how instance lines are referred
to like "Dds" but in a product name, the vcpu count interrupts, so you
can't copy and paste.

AWS is of the form:

    ${memory-ratio:m}${microarchish:5}${manufacturer:a}.${t-shirt-size}

Example: `m5a.xlarge`

== The Sort Assist Byte Range

The general idea here is to choose bytes that are:

* Contiguous (easy to write a regexp range against)
* Meaningless

So I wrote a table of the alphabet and assigned meaning to characters.
The range `[t-w]` stood out as a good range in the lack of meaning
present:

    a - arm, amd
    b
    c - core
    d - disk
    e
    f
    g - gigabyte
    h
    i - intel, instance
    j
    k
    l - thin when in non-monospace fonts
    m - memory, megabyte
    n - network
    o
    p
    q
    r - ram
    s - storage
    t
    u
    v
    w
    x - multiply?
    y
    z

### CODE REVIEW

**Comment by byucesoy** [comment]
I think naming format looks good. I have 3 general comments;

- I'm not sure if we really want to put the year explicitly. Name is coupled with VM types/SKU which is coupled with billing meters. That means every year we would create a new VM types and billing meters. It would be also difficult to perform background upgrades as new CPUs are released. If people explicitly requested amd19 in the past we cannot move them amd20. However if we abstract year information we can change the underlying hardware as we upgrade the fleet. Finally, I don't think many people care about the year of the CPU released, but that is my impression, which might be wrong.
- Adding "meaningless" sort bit seems strange to me. I don't have better alternative unfortunately. We can ignore sorting I guess.
- Would we want to have CPU optimized product line in the future? Will that only be specified with higher core/memory ratio? In Azure (and potentially in other providers as well), apart from the ratio, speed of the each core also increases in CPU optimized families, which is specified in the name (F series for Azure) for example. Do we need something like that?

**Comment by fdr** [comment]
> I think naming format looks good. I have 3 general comments;
> 
>     * I'm not sure if we really want to put the year explicitly. Name is coupled with VM types/SKU which is coupled with billing meters. That means every year we would create a new VM types and billing meters. It would be also difficult to perform background upgrades as new CPUs are released. If people explicitly requested amd19 in the past we cannot move them amd20. However if we abstract year information we can change the underlying hardware as we upgrade the fleet. Finally, I don't think many people care about the year of the CPU released, but that is my impression, which might be wrong.

Maybe for the worse, or perhaps not applicable to us, this is kinda how it works on the other ones, too: you never get a m6a (zen3, amd20) if you were buying m5a, (zen2, amd19) you never get a Dsv4 if you were asking for Dsv5. There are some special meta-product lines like "e2" on GCP. It's also not clear what "t3" on AWS or B on Azure are made out of, maybe they're meta-product as well.

>     * Adding "meaningless" sort bit seems strange to me. I don't have better alternative unfortunately. We can ignore sorting I guess.

It is strange. At the same time, people tolerate a lot of meaningless (to them) characters in product codes.

>     * Would we want to have CPU optimized product line in the future? Will that only be specified with higher core/memory ratio? In Azure (and potentially in other providers as well), apart from the ratio, speed of the each core also increases in CPU optimized families, which is specified in the name (F series for Azure) for example. Do we need something like that?

The ratio is taken care of by encoding RAM into the name, but clock frequency or other intra-manufacturer variation is not. The reason I've punted on this is because the largest vendors tend to not ship multiple processors of one generation, e.g. Hetzner you are on AX161 (until they, one day, ship a Zen3 AX162, presumed), on OVH we'd be on a Scale-7 probably, with few other amd20 options. I suppose we could do an azure-like thing and add modifiers, like "amd20f" for "fast". But this is theoretical at this time.

This also properly levels the two, since OVH will cost more in part from the newer micro-architecture.

One difference we have with the usual suspects is people will need to compare different hardware endowments across providers quickly. One option is we can be even more prescriptive ("standard-1") and hope that our choices of "whatever" are close enough: zen2 on hetzner, zen3 on OVH. The option I chose here is to be more descriptive, which has pros and cons. There are probably some intermediate options.

**Comment by byucesoy** [comment]
> Maybe for the worse, or perhaps not applicable to us, this is kinda how it works on the other ones, too: you never get a m6a (zen3, amd20) if you were buying m5a, (zen2, amd19) you never get a Dsv4 if you were asking for Dsv5.

Yep, but they also don't release new type each year. Also they can run on different CPUs. For example Dv4 can be backed by 8370C (Ice Lake) or 8272CL (Cascade Lake). I guess we can also back amd19 with any AMD CPU released at 2019.

Do you think people would have negative first impression seeing that we run on 4 years old hardware?

**Comment by fdr** [comment]
> > Maybe for the worse, or perhaps not applicable to us, this is kinda how it works on the other ones, too: you never get a m6a (zen3, amd20) if you were buying m5a, (zen2, amd19) you never get a Dsv4 if you were asking for Dsv5.
> 
> Yep, but they also don't release new type each year. Also they can run on different CPUs. For example Dv4 can be backed by 8370C (Ice Lake) or 8272CL (Cascade Lake). I guess we can also back amd19 with any AMD CPU released at 2019.

That's okay, our numbers don't have to be dense. It's weird that Azure ships with two processor two years apart as Dv4 without differentiation, IMO. If we really thought they were close enough and were in the same situation, we'd probably call all of it "intel19".

> Do you think people would have negative first impression seeing that we run on 4 years old hardware?

Maybe, even if it's what m5a is made out of, still the latest and greatest in many regions on AWS. The difference in problem for us is that different providers are going to have different endowments for extensive periods, we'll never bring them into alignment as the integrated hyperscalers tend to do.

**Comment by fdr** [comment]
I'm open to counter-proposal, there are many aspects of this I find unsatisfying. The one thing I like is that I get the RAM without running off to the manual. Crunchy I think it well, _for a database product_: standard-32 has 32 gigabytes of ram. But, this is highly relevant to caching, so, makes sense....

**Comment by fdr** [comment]
I'll lodge my own counter-proposal: do it like Heroku, which is "standard-1." These strings I have here can be useful for compact *description*, but perhaps we can be *prescriptive* on each platform on what is "standard" in terms of memory ratio and architecture. They can also be dynamic, e.g. if you opt into "standard-1" with the understanding you'll receive a microarchitecture update at any time it is convenient.

The rough analog in another domain is "target triplet" seen in the GNU build system and beyond. You don't always use the entire triplet, which is like `x86_64-linux-gnu`, but it's there.

**Comment by fdr** [comment]
On AWS, "m" and on GCP "standard" is 4:1 GB/Core ratio (but, notably, GCP disables SMT so you get one exposed vcpu, and AWS leaves it on, so you get two). I think could do something similar.

4:1 is also hetzner and ovh's default (lowest-spec)...which being *lowest* spec means we can't ship a thing similar to AWS's "c" instances, which are "m" or "r" but with less memory. But that is alright, just an interesting fact.

**Comment by byucesoy** [comment]
We discussed this a bit in EU daily sync as well. Here are few inputs;

- Additional sorting bit that is different for 1/2/3 digit numbers might confuse users even more and it is difficult to remember, thus they would probably need to check the documentation again to learn correct name of the VM they want to use (for API users)
- Having memory/cpu amount in the name is indeed useful.
- Heroku style standard-1 naming is more friendly, when we make our naming too technical we are optimizing ourselves for very technical users, which is probably minority.
- If CPU count is present, all providers use vCPU count instead of CPU count. We would probably want to do the same.

So how about this format;
Standard-2x8 or Standard-2-by-8

In the future, it can be extended to things like MemoryOptimized-2x16. If we want we can also add modifiers at the end, such as;

Standard-2x8-a: AMD processors
Standard-2x8-d: Bigger local disk

**Comment by fdr** [comment]
so, here's question, so we don't end up in azure-land: what is that "standard" string doing? Is it ever not standard?

I think sacrificing sort is fine.

I think we should almost certainly not count by vCPU, because the increasing variation of whether to enable SMT (mostly from ARM, sometimes by election for HPC-style workloads) messes up sensible comparison. Instead, count by the core.

**Comment by byucesoy** [comment]
Maybe GeneralPurpose instead of Standard then, which might be even more in Azure land 😀 but at least more meaningful than Standard. I'm open for anything else as well. I put Standard as "default option you would pick if you don't have specific needs".

On vCores vs. Cores; my concern is that people would compare our core numbers to hyperscalers' vCore numbers. That same path dependency probably applies to us as well. How about using vCores for the SMT enabled setups and cores for the ones without SMT with the cost of some inconsistency for ourselves? Also how does hyperscalers handle the difference due to lack of SMT in ARM?

**Comment by fdr** [comment]
I don't think should worry about the vcpu vs core number thing in comparison with hyperscalers, I think it's a sufficiently theoretical disadvantage that, instead, let's do what is most helpful to the customer to make comparisons. That is cores, I think. It even leaks in that it's the smallest unit of m6a you can buy is two vcpus. And in AWS's case, the number of cores is not even in the model name: `m6a.large`

There is a caveat here: suppose people really do care about knowing how many vcpus are exposed to the: then in-banding vcpu information is better. But somehow, I doubt they do.

The way GCP has to handle this difference is with a asterisk and explaining stuff for their `t2d` vs `n2d`. Cores are easily the most consistent unit in terms of performance and cost.

The situation with "standard" is dissatisfying if we also add more in-band information. let me think about it a bit more. I'll write something.

Basically, I don't think the hyperscalers handle this optimally, a lot of it comes from growing up in an era where x86 with SMT two-way was the only game in town.

**Comment by byucesoy** [comment]
BTW, I'm planning to merge https://github.com/ubicloud/ubicloud/pull/330 after this, because my PR needs a parser and it can either implement one according to old product names (m5a.2x) or wait for the parser in this one.

**Comment by byucesoy** [comment]
> In older days, I didn't know CPU and vCPU in vm sizes. I looked Digital Ocean, AWS, GCP for 2 vcpus general purpose types and choose cheapest one for my side project.

That was my thinking as well. I don't think people really care about knowing how many vcpus are exposed to their OS. Instead the problem is, I suspect lots of people don't know or don't notice that hyperscalers use vCore instead of core. At least I didn't know while I was a junior engineer.

**Comment by ozgune** [comment]
I agree with Burak and Enes in that vCPUs have become the cloud standard. If we model our product names using physical core counts, from the customers' perspective, we're going to lose our price advantage. 95%+ of customers will see the core count in our name and compare our price to the respective price on AWS, DO, Fly, or Hetzner for the vCPU count. They will then decide we're only slightly cheaper.

We're operating in a market where even Microsoft doesn't try to change the market perception. By striving for future technical consistency for ARM family, we'd either (a) need to educate an established $300B/yr market or (b) filter out most customers who'd want to try out Ubicloud. (a) isn't realistic and (b) isn't desirable.

Could we use vCPUs in our product names instead?

Minor note: I also looked at Hetzner's ARM VMs and they also advertise vCPUs. So even our underlying provider for the specific CPU family doesn't use our proposed terminology.

**Comment by fdr** [comment]
I don't really agree with this analysis, it seems like people looked at some data tables that focus on vcpu (which we could do also) and not the actual model names, but it's water under the bridge. For now.

It does screw up....a lot. however.
